File: core/src/main/java/no/nav/pensjon/vtp/felles/OidcTokenGenerator.java
package no.nav.pensjon.vtp.felles;

import static java.util.Collections.singletonList;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.jose4j.jwt.JwtClaims;
import org.jose4j.jwt.NumericDate;
import org.jose4j.lang.JoseException;

import com.google.common.base.Strings;


public class OidcTokenGenerator {

    private final KeyStoreTool keyStoreTool;
    private List<String> aud = singletonList("OIDC");
    private NumericDate expiration = NumericDate.fromSeconds(NumericDate.now().getValue() + 3600*6);
    private String issuer;
    private NumericDate issuedAt = NumericDate.now();
    private final String subject;
    private String nonce;
    private Map<String, String> additionalClaims = new HashMap<>();

    public OidcTokenGenerator(KeyStoreTool keyStoreTool, String brukerId, String nonce) {
        this.keyStoreTool = keyStoreTool;
        additionalClaims.put("azp", "OIDC");
        additionalClaims.put("acr", "Level4");
        this.subject = brukerId;
        this.nonce = nonce;

    }

    public void addAud(String e) {
        this.aud = new ArrayList<>(aud);
        this.aud.add(e);
    }

    OidcTokenGenerator withoutAzp() {
        additionalClaims.remove("azp");
        return this;
    }

    OidcTokenGenerator withExpiration(NumericDate expiration) {
        this.expiration = expiration;
        return this;
    }

    public OidcTokenGenerator withIssuer(String issuer) {
        this.issuer = issuer;
        return this;
    }

    public String create() {
        JwtClaims claims = new JwtClaims();
        claims.setIssuer(issuer);
        claims.setExpirationTime(expiration);
        claims.setGeneratedJwtId();
        claims.setIssuedAt(issuedAt);
        claims.setSubject(subject);
        if (!Strings.isNullOrEmpty(nonce)) {
            claims.setClaim("nonce", nonce);
        }
        if (aud.size() == 1) {
            claims.setAudience(aud.get(0));
        } else {
            claims.setAudience(aud);
        }
        for (Map.Entry<String, String> entry : additionalClaims.entrySet()) {
            claims.setStringClaim(entry.getKey(), entry.getValue());
        }
        try {
            return keyStoreTool.createRS256Token(claims.toJson()).getCompactSerialization();
        } catch (JoseException e) {
            throw new RuntimeException(e);
        }
    }
}

File: mocks/auth-mock/src/main/java/no/nav/pensjon/vtp/auth/OidcTokenGenerator.kt
package no.nav.pensjon.vtp.auth

import com.google.common.base.Strings
import org.jose4j.jwt.JwtClaims
import org.jose4j.jwt.NumericDate
import org.jose4j.jwt.NumericDate.fromSeconds
import org.jose4j.jwt.NumericDate.now
import org.jose4j.lang.JoseException
import java.util.*

class OidcTokenGenerator(
        private val jsonWebKeySupport: JsonWebKeySupport,
        private val subject: String,
        private val nonce: String?,
        private val issuer: String,
        private val aud: MutableList<String> = ArrayList(listOf("OIDC")),
        private var expiration: NumericDate = fromSeconds(now().value + 3600 * 6),
        private val issuedAt: NumericDate = now(),
        private val additionalClaims: Map<String, String> = mapOf(
                "azp" to "OIDC",
                "acr" to "Level4"
        )
) {
    fun addAud(e: String) {
        aud.add(e)
    }

    fun create(): String {
        val claims = JwtClaims()
        claims.issuer = issuer
        claims.expirationTime = expiration
        claims.setGeneratedJwtId()
        claims.issuedAt = issuedAt
        claims.subject = subject
        if (!Strings.isNullOrEmpty(nonce)) {
            claims.setClaim("nonce", nonce)
        }
        if (aud.size == 1) {
            claims.setAudience(aud[0])
        } else {
            claims.audience = aud
        }
        for ((key, value) in additionalClaims) {
            claims.setStringClaim(key, value)
        }
        return try {
            jsonWebKeySupport.createRS256Token(claims.toJson()).compactSerialization
        } catch (e: JoseException) {
            throw RuntimeException(e)
        }
    }
}
